### ISSUE TITLE
Gradual migration from Healpix C++ to ducc back-end

### ISSUE THREAD

**Issue by mreineck** [healpy/healpy#717]
I propose to switch from the current Healpix C++ backend to the `ducc0` package (https://pypi.org/project/ducc0), because of the following advantages
- `ducc0`can be directly installed as a binary wheel (Linux, MacOS, Windows), so the `healpy` installation process should become much easier, and perhaps work out of the box on Windows.
- the code base is newer, and some algorithms are more efficient (e.g. a_lm rotation)
- maybe `cython` will be no longer needed in `healpy`

This is of course a big task, so there are a few important questions:
- is it currently a good time to start this, or should I wait for some milestone (e.g. next `healpy` release)?
- are all main developers (@zonca, @lpsinger, @hivon, @DanielLenz, @crosset) OK with such a change?
- are there any volunteers to help me with this? :-)

My original idea was to work on a branch in my own `healpy` fork, but if more people are interested in working on this, perhaps we should do it directly in the main repository. Any thoughts?

**Comment by hivon** [comment]
Hi Martin,

is this replacement only for healpy backend, or also for the standalone 
Healpix/C++ ?

Cheers,
Eric

**Comment by mreineck** [comment]
For the moment I'd propose it only for the healpy backend since the benefits are mostly (but not entirely) due to the simplier interface and easier packaging. For the standalone C++ version I suggest to adopt only the performance-relevant parts, notably the a_lm rotation code.
There  is one thing that will become more difficult in the future: I do not see a practical way to integrate new algorithmic tweaks to the SHT algorithms into `libsharp` any more, since they just become too convoluted to express in plain C. (Same for the FFT and `libfftpack`, but FFT performance is not that crucial for Healpix.)  At some point we probably have to experiment with linking a C++ support library to Fortran Healpix for the SHTs, and that will probably not be fun...

**Comment by zonca** [comment]
It seems a really good way forward for healpy, now it is a good time because there are no changes since the last release.
I can help a little bit, definitely reviewing pull requests.
Yes, better work in a branch in the healpy repository.

**Comment by mreineck** [comment]
OK, great! I'll probably do this on the side, so it might take a few months.
Technical question: which (internal) interfaces am I allowed to break? Is everything under `src/` considered an implementation detail or are there users (outside `healpy` itself) accessing this?

Some of the pixel related functions might be a bit of a headache, I notice: `healpy` allows specifying the `nside` parameter as an array, but my functions expect it to be a single constant for every invocation. I'll probably have to add another layer of glue code for those functions.

**Comment by mreineck** [comment]
Any suggestion how to proceed, @zonca ?

(I'm not in a hurry, I just want to avoid that we are accidentally waiting for each other :-)

**Comment by zonca** [comment]
no sorry, I have a long backlog of things to work on, I hope to get to this next week

**Comment by mreineck** [comment]
New update: `ducc` now has new code to run SHTs on multiple maps/a_lm simultaneously, which can save time at low resolutions. Also, least-square map analysis is now supported directly in C++.

I wonder if we should aim for a gradual transition from Healpix C++ to `ducc` ... this would mean additional dependencies during the transition period, but on the other hand all changes would get immediate real-life testing and users will not be confronted with a single new release that has a high potential to break really badly.

**Comment by lpsinger** [comment]
Would you be willing to dual-license ducc under BSD? That way we could join forces with astropy-healpix.

**Comment by mreineck** [comment]
Sorry, the reasoning behind the licensing has not changed. The FFT part is available under 3-clause BSD, but the rest will stay GPLv2+ only.

**Comment by mreineck** [comment]
I see that a lot of effort is currently going towards making the `healpy` wheel smaller and reducing its dependencies, and I just wanted to point out that much of this work could have been avoided by a switch to `ducc`, and the time invested could have brought us a large part of the way towards that goal.

If the switch is still on the table, I suggest to get it underway soon. That is to say, I'm happy to do as much implementation work on this as I can, but I need @zonca  to decide on potential interface changes and how much migration pain we can inflict on `healpy` users.

**Comment by zonca** [comment]
I think we should start working on the `ducc` version targeting a `healpy` 2.0.0 release, taking the opportunity to restructure everything, no worries about backward compatibility.
And we will continue to also maintain `healpy` 1.

**Comment by zonca** [comment]
I think we should also change repository, what about creating a `healpy2` repository and start development there?
so we could keep issues and pull requests separate for the 2 versions.
We will continue to maintain `healpy 1.x` for a few years at least in parallel with development on healpy 2.

@lpsinger do you think it is better to keep 1 repository only?

**Comment by mreineck** [comment]
> I think we should also change repository, what about creating a healpy2 repository and start development there?

From the implementer's perspective that would certainly be the best scenario by far!
And I think the advantages would also be compelling enough for users to switch, at least when they start new projects.

**Comment by mreineck** [comment]
From @lpsinger's perspective it's probably a bit of a nightmare scenario, since he will most likely have not one, but two `healpy` wheels in future deployments ...

**Comment by hivon** [comment]
@mreineck: how much the SHT or random generated results from ducc0 numerically differ from those of Healpix/C++ (or libsharp) for their respective default settings ?
and will  some settings  allow *scientific* compatibility between them (forgetting the interface compatibility).
About interfaces: a few years ago you sounded pessimistic about the interface compatibility of, for instance, ang2pix, but more optimistic recently thanks to a more flexible ducc0. What is your take now ?

**Comment by mreineck** [comment]
> @mreineck: how much the SHT or random generated results from ducc0 numerically differ from those of Healpix/C++ (or libsharp) for their respective default settings ?

They will differ by numerical noise only, since the underlying algorithms are the same. `ducc0`'s SHT algorithm was generated as a copy from the `libsharp` one after `libsharp` switched to the Ishioka recurrence, and since that time no significant improvements have happened.

The functions in `ducc` don't really depend on any default settings. There is `synthesis` (equivalent to `alm2map`) that goes from `alm` with a given band limit to a map with clearly defined pixelisation (Healpix, GL, fully general, etc.), and there is `adjoint_synthesis` which performs the adjoint operation within numerical accuracy,
Note that there is no `map2alm` operation involving ring or pixel weights, or potential iteration. If any of these are desired, they have to be implemented on the `healpy` side of things.

> and will some settings allow *scientific* compatibility between them (forgetting the interface compatibility).

I do not think that any special settings will be necssary to achieve this.

(There is one SHT-related thing that I have considered changing, and that's the conventional minus sign in front of E and B a_lm for `spin!=0`; it really confuses people outside astronomy. However, at the moment this is still in the code, and if it ever changes I will make absolutely sure that users will notice the change and can act accordingly.)

> About interfaces: a few years ago you sounded pessimistic about the interface compatibility of, for instance, `ang2pix`, but more optimistic recently thanks to a more flexible ducc0. What is your take now ?

I have come closer, but (using `ang2pix` as example) fully matching the current `healpy` interface will still be inefficient for two reasons:
- `ducc` wants to have `theta` and `phi` provided in a single array instead of two arrays. This can be fixed by copying the angles together on the Python side and is only a minor problem
- `ducc` assumes that `nside` will be identical for all inputs of a single call, which is a clear difference in functionality.

Both of these problems could be overcome by writing a C++ "glue module" which implements all these functions in the way `healpy` would like them to be, but @zonca is not a big fan of this, which I can understand very well. Another alternative (putting all of these functions directly into `ducc`) is something I'd prefer to avoid, since it could become quite a maintenance burden for me.
So *if* we can go for new, more minimalistic interfaces, I'm all for it.

**Comment by mreineck** [comment]
@hivon: I just realized that by "default settings" you probably mean "which lmax to use by default when analyzing a map with a given nside". These are all things that have to be decided inside the wrapper code (and can be made totally compatible to `healpy1`); `ducc` just does dumb calculations, it never tries to second-guess what the user could have wanted.

**Comment by hivon** [comment]
I was also thinking about the number of iterations, use of weights, ..., criterion to deem an alm satisfactory (I think ducc has several possible targets).
How about the random generation part: should we leave it to the user (what is implied in healpy documentation does not agree with numpy recommendations) or try to provide a more foolproof mechanism (that may go beyond a different starting seed for each simulation), especially if we aim at massively parallel simulations.

**Comment by mreineck** [comment]
> I was also thinking about the number of iterations, use of weights, ..., criterion to deem an alm satisfactory (I think ducc has several possible targets).

This is all outside `ducc`'s scope and needs to be done by hand in the wrapper.
(There is an iterative inversion routine provided for convenience, but that requires specification of a desired residual error. If backwards compatibility is desired, this cannot be used.)
> How about the random generation part: should we leave it to the user (what is implied in healpy documentation does not agree with numpy recommendations) or try to provide a more foolproof mechanism (that may go beyond a different starting seed for each simulation), especially if we aim at massively parallel simulations.

These are all important questions, but also outside the `ducc` scope. It will not generate random numbers, so functions like `syn_alm` have to be implemented outside, which is luckily not performance critical.

**Comment by mreineck** [comment]
This is also the reason why the 1/f-noise functionality in `ducc0.misc` requires that a stream of random numbers is provided as input.

**Comment by lpsinger** [comment]
> From @lpsinger's perspective it's probably a bit of a nightmare scenario, since he will most likely have not one, but two `healpy` wheels in future deployments ...

So. This is very timely because I have been thinking about the broader HEALPix in Python landscape. I am pressed for time today because I am packing for my son's Scout troop's caving trip this weekend. I'll try to write at length about my ideas this coming week.

**Comment by zonca** [comment]
we should also take a look at interfaces provided by other packages:

* https://astropy-healpix.readthedocs.io/en/latest/
* https://mhealpy.readthedocs.io/en/latest/index.html

**Comment by lpsinger** [comment]
I am the maintainer of astropy-healpix, and one of my colleagues at Goddard maintains mhealpy.

**Comment by ntessore** [comment]
Apart from these two, there's [the "healpix" package for Python](https://github.com/ntessore/healpix), which I maintain. It's a very minimalistic numpy vectorisation of the BSD-licensed `healpix_bare` library. Mostly, care has been taken that operations do not create intermediary arrays. In addition, the package allows sub-pixel indexing, random point picking, and some initial support for MOC indexing schemes. The interface has been kept healpy-compatible as much as possible (see [WIP docs](https://healpix--36.org.readthedocs.build/36/healpix.html)).
